### PR TITLE
Parse artifact path correctly in windows UNC format with servername

### PR DESCRIPTION
**Pull request by bali0019**
Signed-off-by: Jas Bali <bali0019@gmail.com>

## Related Issues/PRs

 #7645

## What changes are proposed in this pull request?

In order to parse the artifact path correctly on windows when it contains a server name/UNC format, this PR explicitly parses out the `netloc` and rebuilds the path component containing that `netloc` in UNC format so that `pathlib` can correctly resolve it.

## How is this patch tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/file_utils.py [hunk 1]
```diff
@@ -534,7 +535,13 @@ def local_file_uri_to_path(uri):
     Convert URI to local filesystem path.
     No-op if the uri does not have the expected scheme.
     """
-    path = urllib.parse.urlparse(uri).path if uri.startswith("file:") else uri
+    path = uri
+    if uri.startswith("file:"):
+        parsed_path = urllib.parse.urlparse(uri)
+        path = parsed_path.path
+        # Fix for retaining server name in UNC path.
+        if platform.system().lower() == "windows" and parsed_path.hostname:
```
Comment (BenWilson2) [inline mlflow/utils/file_utils.py:543]: you added a great utility to `test.helper_functions` in the form of `is_local_os_windows` but the same logic is used here. Can that test helper just be moved to MLflow utilities and have the tests import it in the test suites?

**Review by BenWilson2 — APPROVED** [review]
Small nit; otherwise LGTM!

File: tests/artifacts/test_artifacts.py [hunk 2]
```diff
@@ -224,3 +226,38 @@ def test_artifact_logging_resolution_works_with_non_root_working_directory(text_
...
+    )
+    with mlflow.start_run(experiment_id=experiment_test_2_id) as run:
+        with mock.patch("shutil.copyfile"), mock.patch("os.path.exists") as mock_unc_path:
+            mock_unc_path.return_value = True
+            mlflow.log_artifact(text_artifact.artifact_path)
+            artifact_uri = mlflow.artifacts.download_artifacts(
+                run_id=run.info.run_id, artifact_path=text_artifact.artifact_name
+            )
+            assert (
+                rf"{local_file_uri_to_path(experiment_test_2_artifact_location)}"
```
Comment (harupy) [inline tests/artifacts/test_artifacts.py:261]:
```suggestion
                local_file_uri_to_path(experiment_test_2_artifact_location)}
```

File: tests/artifacts/test_artifacts.py [hunk 3]
```diff
@@ -224,3 +226,38 @@ def test_artifact_logging_resolution_works_with_non_root_working_directory(text_
...
+
+    experiment_test_2_artifact_location = "file://my_server/my_path/my_sub_path"
+    experiment_test_2_id = mlflow.create_experiment(
+        "test_exp_e", experiment_test_2_artifact_location
+    )
+    with mlflow.start_run(experiment_id=experiment_test_2_id) as run:
+        with mock.patch("shutil.copyfile"), mock.patch("os.path.exists") as mock_unc_path:
+            mock_unc_path.return_value = True
+            mlflow.log_artifact(text_artifact.artifact_path)
+            artifact_uri = mlflow.artifacts.download_artifacts(
```
Comment (harupy) [inline tests/artifacts/test_artifacts.py:257]:
```suggestion
            local_path = mlflow.artifacts.download_artifacts(
```

File: tests/artifacts/test_artifacts.py [hunk 4]
```diff
@@ -224,3 +226,38 @@ def test_artifact_logging_resolution_works_with_non_root_working_directory(text_
...
+
+
+@pytest.mark.skipif(not is_local_os_windows(), reason="This test only passes on Windows")
+def test_log_artifact_windows_path_with_hostname(text_artifact):
+    experiment_test_1_artifact_location = r"\\my_server\my_path\my_sub_path\1"
+    experiment_test_1_id = mlflow.create_experiment(
+        "test_exp_d", experiment_test_1_artifact_location
+    )
+    with mlflow.start_run(experiment_id=experiment_test_1_id) as run:
+        with mock.patch("shutil.copyfile"), mock.patch("os.path.exists") as mock_unc_path:
```
Comment (harupy) [inline tests/artifacts/test_artifacts.py:238]: Why do we need to patch `shutil.copyfile` and `os.path.exists`?

File: tests/artifacts/test_artifacts.py [hunk 5]
```diff
@@ -224,3 +226,38 @@ def test_artifact_logging_resolution_works_with_non_root_working_directory(text_
             run.info.run_id,
         )
     os.chdir(original_cwd)
+
+
+@pytest.mark.skipif(not is_local_os_windows(), reason="This test only passes on Windows")
+def test_log_artifact_windows_path_with_hostname(text_artifact):
+    experiment_test_1_artifact_location = r"\\my_server\my_path\my_sub_path\1"
```
Comment (harupy) [inline tests/artifacts/test_artifacts.py:233]:
```suggestion
    experiment_test_1_artifact_location = r"\my_server\my_path\my_sub_path\1"
```

File: mlflow/utils/helper_functions.py [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import platform
```
Comment (harupy) [inline mlflow/utils/helper_functions.py:1]: Let's rename this to `os.py`

File: mlflow/utils/helper_functions.py [hunk 7]
```diff
@@ -0,0 +1,8 @@
...
+
+
+def is_local_os_windows():
```
Comment (harupy) [inline mlflow/utils/helper_functions.py:4]:
```suggestion
def is_windows():
```

Can we rename this function to `is_windows`?

File: tests/artifacts/test_artifacts.py [hunk 8]
```diff
@@ -224,3 +226,38 @@ def test_artifact_logging_resolution_works_with_non_root_working_directory(text_
...
+    experiment_test_1_id = mlflow.create_experiment(
+        "test_exp_d", experiment_test_1_artifact_location
+    )
+    with mlflow.start_run(experiment_id=experiment_test_1_id) as run:
+        with mock.patch("shutil.copyfile"), mock.patch("os.path.exists") as mock_unc_path:
+            mock_unc_path.return_value = True
+            mlflow.log_artifact(text_artifact.artifact_path)
+            artifact_uri = mlflow.artifacts.download_artifacts(
```
Comment (harupy) [inline tests/artifacts/test_artifacts.py:241]:
```suggestion
            local_path = mlflow.artifacts.download_artifacts(
```

File: tests/artifacts/test_artifacts.py [hunk 9]
```diff
@@ -224,3 +226,38 @@ def test_artifact_logging_resolution_works_with_non_root_working_directory(text_
...
+                rf"\artifacts\{text_artifact.artifact_name}" == artifact_uri
+            )
+
+    experiment_test_2_artifact_location = "file://my_server/my_path/my_sub_path"
+    experiment_test_2_id = mlflow.create_experiment(
+        "test_exp_e", experiment_test_2_artifact_location
+    )
+    with mlflow.start_run(experiment_id=experiment_test_2_id) as run:
+        with mock.patch("shutil.copyfile"), mock.patch("os.path.exists") as mock_unc_path:
+            mock_unc_path.return_value = True
```
Comment (harupy) [inline tests/artifacts/test_artifacts.py:255]: We can use `return_value` when calling `mock.patch`:

```python
with mock.patch(..., return_value=True):
   ...
```

File: tests/artifacts/test_artifacts.py (hunk 4 above)
Comment (bali0019) [reply tests/artifacts/test_artifacts.py]: For this test case, I need to test a UNC path with the server name/network drive part. So, I just need to [mock this part](https://github.com/mlflow/mlflow/blob/master/mlflow/store/artifact/local_artifact_repo.py#L36) that checks if the path (in this case on a network drive) exists ...if it does then it copies that local file to network drive path [using this] (https://github.com/mlflow/mlflow/blob/master/mlflow/store/artifact/local_artifact_repo.py#L38)

File: tests/artifacts/test_artifacts.py (hunk 5 above)
Comment (bali0019) [reply tests/artifacts/test_artifacts.py:233]: Yes, just so that on windows, we can force the network drive/server name to be part of the absolute path.

Comment (harupy) [reply tests/artifacts/test_artifacts.py:233]: Understood :)

File: tests/artifacts/test_artifacts.py (hunk 4 above)
Comment (harupy) [reply tests/artifacts/test_artifacts.py]: Does `shutil.copyfile` and `os.path.exists` work for a UNC path?

File: mlflow/utils/os.py [hunk 10]
```diff
@@ -0,0 +1,8 @@
+import platform
+
+
+def is_windows():
+    """
+    :return: Returns true if the local system/OS name is Windows.
+    """
+    return platform.system().lower() == "windows"
```
Comment (dbczumar) [inline mlflow/utils/os.py:8]: I don't think this is a reliable check for all Windows systems.

According to https://stackoverflow.com/questions/1325581/how-do-i-check-if-im-running-on-windows-in-python, some Windows systems might use `win32` or `win64`.

Can we use `os.name == 'nt'` instead?

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @bali0019 !

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**